### ISSUE TITLE
Time committing a lot of xml

### ISSUE THREAD

**Issue by JDSBCC** [ppazos/cabolabs-ehrserver#360]
Hi,
Im doing a lot of tests in my application and I noticed that committing xml takes to long. So as I said in the other issue, I'm using threads, but it doesn't makes difference because when I start to commit the xml all threads needs to wait until commit is finished. The request itself is really fast but the data is not inserted so fast in the database. Im just doing a test with 3676 documents but my database has more than 1 million. Still, with 3676 documents I spent 5h. Just 30min for my application (with all request to EHR except commit xml) and 4:30 just for committing all xml.
Do you think I'm doing something wrong, or is it normal?
Is there a way to speed up this process?

Thanks

**Comment by ppazos** [comment]
@JDSBCC it really depends on the server hardware. How many commits in parallel can be handled by the server would depend on the number of CPUs available, processing speed might depend on CPU speed and available RAM memory, and also on disk speed.

I think the XML processing can be optimized a little, but it is doing a lot of validations to be sure the received XML is correct and tries to give meaningful responses so if an error is detected you can fix your client apps.

In a not-so-big server in the Open Shift cloud (https://cabolabs-ehrserver.rhcloud.com/ehr) I would say I can commit 1 composition per second manually, that's 3600 commits per hour. With an automatic committed this can be improved a little, and of course, with a big dedicated server (lot of cores and memory and SSD) I would say this performance can be improved in many orders of magnitude.

Just for the record, the EHRServer doesn't have any constraints about quantities of things you can commit or send through the API. The limits are imposed by 1. the web server, 2. the database server, 3. the hardware.

Hope that helps!

**Comment by JDSBCC** [comment]
Yes, it helps =). I will test with another machine and I hope this process becomes faster. If I got something like 3h its ok for me. Thank you

**Comment by ppazos** [comment]
@JDSBCC for this kind of use I would recommend to setup EHRServer on a machine with moderated hardware by today's standards (4 cores, 8 GB RAM, SSD) and turn off all the non-esencial software / processes running on that machine (at least while you are working loading data on the EHRServer). Other software running on the machine can also make the EHRServer slower.

Another think you can try, but I can't guarantee it will improve performance, is to have the DB server installed on a different machine than the web server:

`client app --> EHRServer (web server) --> EHRServer (db server)`

**Comment by JDSBCC** [comment]
Hi, you were right. I tested my application with another machine (core i7, 8Gb with no ssd) and I got 55min. I also made a few changes in my app (i removed threads because shared areas had many problems with them). I dont have time to test the implementation of servers in different machines because I have to conclude this until June 24 (Its a thesis for university), but I really appreciate all your help. Thank you =)
